### ISSUE TITLE
suggestion add `is_confidential` to oauth2 base.Client

### ISSUE THREAD

**Issue by jvanasco** [oauthlib/oauthlib#597]
I discovered this as part of #585, but it's a separate (though related) issue...

The base `Client` object does not track if a client is confidential (or not), however clients must act differently in certain situations depending on this.  It looks like this was largely handled by creating a different base class for each 'concept'.  I think adding an `is_confidential' property to the base `Client` class (and set in the subclasses) would allow for more code re-use and just be generally more descriptive and self-documenting.

**Comment by skion** [comment]
If we do this, would this void or simplify #593 to a large extend? If so, then it might be the better direction in the first place?

**Comment by jvanasco** [comment]
It won't void #593, but I think it would simplify future maintenance and integration with requests_oauthlib. 

It looks like several sections of the two libraries are subclassing functions or altering the behavior based on detecting the class type (e.g. WebApplicationClient vs MobileApplicationClient) where the behavior is actually dependent on whether the client is "confidential" or "public".

More importantly, a client being "confidential" or not is an attribute/descriptor of clients in the oAuth2 spec - but it's not explicitly part of the oauthlib clients.

**Comment by jvanasco** [comment]
I did a deepdive into the code. After consideration, it seems like this request (#597) would simply be a method of documenting the type of client at this time.  It certainly can wait until after 585/593, which (IMHO) should be handled sooner than later.

While the various subclasses of `Client` alter their performance based on this attribute, they are also invoking slightly different arguments and endpoints.  I don't think there is any potential simplification of code at this point.

**Comment by thedrow** [comment]
Is this issue still actionable or should we close it?

**Comment by jvanasco** [comment]
I think it's still actionable.

looking at the current master, and oauth2 only, the is_confidential attribute -- based on public/confidential distinction in the rfc - would be the following

base.py Client
	is_confidential = None

backend_application.py BackendApplicationClient
	is_confidential = False
	
legacy_application LegacyApplicationClient
	is_confidential = False

mobile_application.py MobileApplicationClient
	is_confidential = False

service_appication.py ServiceApplicationClient
	is_confidential = False

web_application.py WebApplicationClient
	is_confidential = True

**Comment by JonathanHuot** [comment]
BackendApplicationClient is always confidential.
MobileApplicationClient is always public.

About the others, they are sometime confidential, sometime public.
I think we can rely on the existence of the client_secret to define if they are confidential or public.
However, it means that's not a class method but an instance method.

Thoughts? 

Legacy aka ROPC:
https://tools.ietf.org/html/rfc6749#section-4.3.2

>    If the client type is confidential or the client was issued client
>    credentials (or assigned other authentication requirements), the
>    client MUST authenticate with the authorization server as described
>    in Section 3.2.1.

Web aka AuthorizationCode:
https://tools.ietf.org/html/rfc6749#section-2.1

>    public
>       Clients incapable of maintaining the confidentiality of their
>       credentials (e.g., clients executing on the device used by the
>       resource owner, such as an installed native application or a web
>       browser-based application), and incapable of secure client
>       authentication via any other means.

Service aka Assertion:
https://tools.ietf.org/html/rfc7521#section-4.1
>    Authentication of the client is optional, as described in
>    Section 3.2.1 of OAuth 2.0 [RFC6749], and consequently, the
>    "client_id" is only needed when a form of client authentication that
>    relies on the parameter is used.

**Comment by jvanasco** [comment]
I based my list off of the current implementations which use the RFC concept of confidential.  The docstrings for most (all?) of these say public or confidential and refer to the rfc’s meaning of the term; for example, WebApplicationClient states it is confidential in the class docstring.

**Comment by skion** [comment]
>  I think we can rely on the existence of the client_secret to define if they are confidential or public.

I agree this is how it should be in practice, but technically it is still possible to have a public client using a `client_secret`. The [best practices](https://tools.ietf.org/html/rfc8252#section-8.5) state the following about this situation:

> Authorization servers that still require a statically included shared secret for native app clients MUST treat the client as a public client[...]

If we do what you suggest, the class would be marked as `confidential` whilst according to the RFC it "MUST" be treated as `public`. I feel that someone might shoot themselves in the foot if we provide such a gun.
